share/jupyter/voila/templates/default/static/main.js

```diff
@@ -40,7 +40,9 @@ require(['static/voila'], function(voila) {
         async function init() {
             // it seems if we attach this to early, it will not be called
             window.addEventListener('beforeunload', function (e) {
-                kernel.shutdown();
+                const xhttp = new XMLHttpRequest();
+                xhttp.open("DELETE", `/api/kernels/${kernel.id}`, false);
```

Comment:
Note that if end up deciding that this is the best way, we would need to ensure we get/set all the appropriate headers and cookies.

Reply:
Which is why using `sendBeacon` instead sounds interesting :+1:  (https://github.com/voila-dashboards/voila/issues/528 for more context)

Reply:
There are more things than just cookies though, i.e. a bunch of headers:
https://github.com/jupyterlab/jupyterlab/blob/64425baacfba677b75efaf05cfb825a82264230d/packages/services/src/serverconnection.ts#L236-L282

How many of those are needed for the DELETE request is a bit unknown though.

I think it might be possible to do a workaround by using a ServerConnection with our own settings that override `fetch` such that it instead uses a sync XHTTPRequest? It would require "translating" the request first though.

Reply:
> There are more things than just cookies though, i.e. a bunch of headers:

Indeed. And these are not picked but by `sendBeacon`.